retrieve-logs/fetch_missing_matomo_requests.py

```diff
@@ -0,0 +1,161 @@
+    try:
+        return int(os.getenv(PERIOD_WIDTH_IN_SECONDS, 60 * 5))
+    except ValueError:
+        get_logger().exception('PERIOD_WIDTH_IN_SECONDS has an invalid format. Please specify an integer.')
+        exit(1)
+
+
+def get_output_filename(start_datetime, end_datetime):
+    filename = os.getenv(OUTPUT_FILENAME)
+    if filename is None or len(filename) < 1:
+        return start_datetime.strftime(DATE_FORMAT) + '_' + end_datetime.strftime(DATE_FORMAT) + FILENAME_SUFFIX
+    return filename
+
+
+def wait_for_the_query_to_complete(response):
+    queryId = response['queryId']
+    status = 'Running'
+    seconds_slept = 0
+    while status != 'Complete':
+        time.sleep(1)
+        seconds_slept += 1
+        if seconds_slept % 30 == 0:
+            get_logger().debug(f'Still waiting for a request. Spent {seconds_slept} seconds waiting so far.')
+        response = client.get_query_results(queryId=queryId)
+        status = response['status']
+    return response
+
+
+def run_query(start_timestamp, end_timestamp):
+    return client.start_query(
+        logGroupName='matomo',
+        startTime=int(start_timestamp.timestamp() * 1000),
+        endTime=int(end_timestamp.timestamp() * 1000),
+        queryString=
+        """fields @message
+        | sort @timestamp asc
+        | filter @logStream like /matomo-nginx/
+        | filter status!='200'
+        | filter status!='204'
+        | filter user_agent!='ELB-HealthChecker/2.0'
+        | filter path like /idsite=1/
+        | filter path like /rec=1/""",
+        limit=10000
+    )
+
+
+def write_requests_to_a_file(response, period_start, period_end, output_filename):
+    count_written = 0
+    with open(output_filename, 'a+') as f:
+        for message in response['results']:
+            if len(message) >= MAX_REQUESTS:
+                log_too_many_requests_and_exit(period_start, period_end)
+            for message in message:
+                if message['field'] == '@message':
+                    f.write(message['value'] + '\n')
+                    count_written += 1
+                    break
+    if count_written > 0:
+        get_logger().debug(
+                f'Wrote {count_written} requests to file {output_filename}'
+                + f' from within the period {period_start} to {period_end}')
+    return count_written
+
+
+if __name__ == '__main__':
+    validate_environment_variables()
+
+    client = boto3.client('logs')
+
+    start_datetime = get_start_datetime()
+    num_of_days = get_number_of_days()
+    period_width = get_period_width()
+    end_datetime = start_datetime + timedelta(days=num_of_days, microseconds=-1)
+
+    output_filename = get_output_filename(start_datetime, end_datetime)
+    if os.path.exists(output_filename):
+        os.remove(output_filename)
+
+    period_start = datetime.utcfromtimestamp(start_datetime.replace(tzinfo=timezone.utc).timestamp())
+    total_written = 0
+    while period_start < end_datetime:
```

Comment:
Should be `<=`??

Reply:
This was fixed in #3 - I just hadn't rebased this branch. I've rebased it now, but that's also messed up the commit order (because GitHub orders by author date, but I made changes to the base branch after I made this one...). Currently, the 5 commits new here are:

- [Remove direct references to environment variables](https://github.com/alphagov/verify-matomo-utils/pull/4/commits/8dbc9527439babc1d2544ad6d1600461bf66b1af)
- [Switch from bash to sh ](https://github.com/alphagov/verify-matomo-utils/pull/4/commits/b44446382a266f61cbfef918335811233cf65851)
- [Switch to Python 3-alpine image](https://github.com/alphagov/verify-matomo-utils/pull/4/commits/4d05da0c47bec16ca8bdba6e5b401d2c92afcdd1)
- [Allow dry runs on replay ](https://github.com/alphagov/verify-matomo-utils/pull/4/commits/94f32f7f5448c2a5485ef427680833f77c83ae56)
- [Allow use of a keyword for specifying start date ](https://github.com/alphagov/verify-matomo-utils/pull/4/commits/a4432e9864ec9b8baeb0ad712a7adafd7b105181)